### PR TITLE
[GH-558] Add support for listing roles

### PR DESCRIPTION
**Pull request by ChandanChainani**
Part:
- https://github.com/mattermost/mmctl/issues/558
- https://github.com/mattermost/mattermost-server/issues/21197

### CODE REVIEW

**Comment by ChandanChainani** [comment]
@isacikgoz Please can you review the MR.

**Comment by isacikgoz** [comment]
@ChandanChainani can you fix the failing CI checks? Let us know if you need any help!

**Comment by ChandanChainani** [comment]
@isacikgoz Fixed please check.

**Comment by isacikgoz** [comment]
@ChandanChainani there is something wrong with the test script, I'll submit a fix for that. Then we will be good to go.

File: commands/permissions_role.go [hunk 1]
```diff
@@ -156,6 +165,19 @@ func showRoleCmdF(c client.Client, cmd *cobra.Command, args []string) error {
...
 }
 
+func listRoleCmdF(c client.Client, cmd *cobra.Command, args []string) error {
+	roles, _, err := c.GetAllRoles()
+	if err != nil {
+		return err
+	}
+
+	for _, role := range roles {
+		printer.PrintT(prettyRole(role), nil)
```
Comment (isacikgoz) [inline commands/permissions_role.go:175]:
```suggestion
		printer.Print(prettyRole(role))
```

**Review by isacikgoz — COMMENTED** [review]
Thanks @ChandanChainani can you also add a e2e test to verify this? Apart from that I have a suggestion to enable JSON outputs.

**Comment by isacikgoz** [comment]
> can you also add a e2e test to verify this? 

@ChandanChainani can you please add a test case for this, thanks!

**Comment by ChandanChainani** [comment]
@isacikgoz Added test. Please review

File: commands/permissions_e2e_test.go [hunk 2]
```diff
@@ -108,3 +108,35 @@ func (s *MmctlE2ETestSuite) TestRemovePermissionsCmd() {
...
+	s.Require().Nil(err)
+	mockRoleNameWithPermissions := "mockrole" + model.NewId()
+	mockRolePermissions := []string{"sysconsole_write_site", "edit_brand"}
+	_, err = s.th.App.CreateRole(&model.Role{Name: mockRoleNameWithPermissions, Permissions: mockRolePermissions})
+	s.Require().Nil(err)
+
+	s.RunForSystemAdminAndLocal("Should list all roles for syasdmin and local clients", func(c client.Client) {
+		printer.Clean()
+
+		err := listTeamsCmdF(c, &cobra.Command{}, []string{})
```
Comment (isacikgoz) [inline commands/permissions_e2e_test.go:125]: I'm confused a bit, @ChandanChainani why `listTeamsCmdF` being used here?

File: commands/permissions_e2e_test.go [hunk 3]
```diff
@@ -108,3 +108,35 @@ func (s *MmctlE2ETestSuite) TestRemovePermissionsCmd() {
 		s.Require().NotContains(updatedRole.Permissions, model.PermissionCreateDirectChannel.Id)
 	})
 }
+
+func (s *MmctlE2ETestSuite) TestListRolesCmdF() {
+	s.SetupTestHelper().InitBasic()
+	mockRoleName := "mockrole" + model.NewId()
+	_, err := s.th.App.CreateRole(&model.Role{Name: mockRoleName})
+	s.Require().Nil(err)
```
Comment (isacikgoz) [inline commands/permissions_e2e_test.go:116]:
```suggestion
	s.Require().NoError(err)
```

File: commands/permissions_e2e_test.go [hunk 4]
```diff
@@ -108,3 +108,35 @@ func (s *MmctlE2ETestSuite) TestRemovePermissionsCmd() {
...
+
+		roleWithPermissions := printer.GetLines()[1].(*model.Role)
+		s.Equal(mockRoleNameWithPermissions, roleWithPermissions.Name)
+	})
+
+	s.Run("Should not list teams for Client", func() {
+		printer.Clean()
+
+		err := listTeamsCmdF(s.th.Client, &cobra.Command{}, []string{})
+		s.Require().Nil(err)
```
Comment (isacikgoz) [inline commands/permissions_e2e_test.go:139]:
```suggestion
		s.Require().NoError(err)
```

File: commands/permission_role_test.go [hunk 5]
```diff
@@ -465,3 +465,50 @@ Permissions   edit_brand
 		s.Require().Len(printer.GetErrorLines(), 0)
 	})
 }
+
+func (s *MmctlUnitTestSuite) TestListRoleCmd() {
+	s.Run("List roles", func() {
+		printer.Clean()
+		printer.SetFormat(printer.FormatPlain)
+		defer printer.SetFormat(printer.FormatJSON)
```
Comment (isacikgoz) [inline commands/permission_role_test.go:473]: Is default format is `JSON`? I don't think we need to set it back to json.

File: commands/permission_role_test.go [hunk 6]
```diff
@@ -465,3 +465,50 @@ Permissions   edit_brand
...
+		}}
+
+		s.client.
+			EXPECT().
+			GetAllRoles().
+			Return(mockRoles, &model.Response{}, nil).
+			Times(1)
+
+		err := listRoleCmdF(s.client, &cobra.Command{}, []string{})
+		s.Require().Nil(err)
```
Comment (isacikgoz) [inline commands/permission_role_test.go:491]:
```suggestion
		s.Require().NoError(err)
```

**Review by isacikgoz — COMMENTED** [review]
Thanks @ChandanChainani I have a couple of questions, and have some minor suggestions.

File: commands/permission_role_test.go (hunk 5 above)
Comment (ChandanChainani) [reply commands/permission_role_test.go:473]: followed similar structure to `TestShowRoleCmd`

**Comment by ChandanChainani** [comment]
@isacikgoz 

Not sure what's the issue in my system it is getting passed today but getting failed tomorrow :rofl: 
```
--- PASS: TestPrintPreparedT (0.00s)
    --- PASS: TestPrintPreparedT/should_execute_template (0.00s)
    --- PASS: TestPrintPreparedT/should_fail_to_execute,_no_method_or_field (0.00s)
=== RUN   TestFlushJSON
=== RUN   TestFlushJSON/should_print_a_line_in_JSON_format
=== RUN   TestFlushJSON/should_print_multi_line_in_JSON_format
--- PASS: TestFlushJSON (0.00s)
    --- PASS: TestFlushJSON/should_print_a_line_in_JSON_format (0.00s)
    --- PASS: TestFlushJSON/should_print_multi_line_in_JSON_format (0.00s)
=== RUN   TestFlushPlain
=== RUN   TestFlushPlain/should_print_a_line_in_plain_format
=== RUN   TestFlushPlain/should_print_multi_line_in_plain_format
=== RUN   TestFlushPlain/should_print_multi_line_in_plain_format_without_a_newline
--- PASS: TestFlushPlain (0.00s)
    --- PASS: TestFlushPlain/should_print_a_line_in_plain_format (0.00s)
    --- PASS: TestFlushPlain/should_print_multi_line_in_plain_format (0.00s)
    --- PASS: TestFlushPlain/should_print_multi_line_in_plain_format_without_a_newline (0.00s)
PASS
ok  	github.com/mattermost/mmctl/v6/printer	(cached)
```

Maybe due to caching I am running below command
```
make test
```

**Comment by isacikgoz** [comment]
@ChandanChainani it caches when there is no change in the code, so it's probably something else. I can look into if you couldn't figure out the issue, just let me know.

**Comment by ChandanChainani** [comment]
@isacikgoz while running e2e test cases with below cmd
```
make test-e2e
```
getting below error
```
==========================\n"}
{"level":"debug","msg":"morph.go:182: ==  threads_threadteamid: migrated (0.0010s)  ======================================================\n"}
{"level":"debug","msg":"Deleting any unused pre-release features"}
panic: Failed to find folder i18n

goroutine 1 [running]:
github.com/mattermost/mattermost-server/v6/testlib.getTestResourcesToSetup()
	$HOME/mmctl/vendor/github.com/mattermost/mattermost-server/v6/testlib/resources.go:92 +0x305
github.com/mattermost/mattermost-server/v6/testlib.SetupTestResources()
	$HOME/mmctl/vendor/github.com/mattermost/mattermost-server/v6/testlib/resources.go:116 +0x2f
github.com/mattermost/mattermost-server/v6/testlib.(*MainHelper).setupResources(0xc000040f60)
	$HOME/mmctl/vendor/github.com/mattermost/mattermost-server/v6/testlib/helper.go:134 +0x2a
github.com/mattermost/mattermost-server/v6/testlib.NewMainHelperWithOptions(0xc00149fe15)
	$HOME/mmctl/vendor/github.com/mattermost/mattermost-server/v6/testlib/helper.go:62 +0x1ec
github.com/mattermost/mmctl/v6/commands.TestMain(0x0?)
	$HOME/mmctl/commands/main_test.go:22 +0x3f
main.main()
	_testmain.go:63 +0x308
FAIL	github.com/mattermost/mmctl/v6/commands	0.701s
testing: warning: no tests to run
PASS
ok  	github.com/mattermost/mmctl/v6/printer	(cached) [no tests to run]
FAIL
make: *** [Makefile:145: test-e2e] Error 1
```

**Comment by hanzei** [comment]
Thank you for your contribution. :+1:  `mmctl` has been moved into https://github.com/mattermost/mattermost/tree/master/server/cmd/mmctl.  Please resubmit your PR to https://github.com/mattermost/mattermost.

**Pull request closed without merging**